source/common/conn_pool/conn_pool_base.cc

```diff
@@ -66,24 +66,25 @@ bool ConnPoolImplBase::shouldCreateNewConnection(float global_preconnect_ratio)
     return pending_streams_.size() > connecting_stream_capacity_;
   }
 
-  // If global preconnecting is on, and this connection is within the global
-  // preconnect limit, preconnect.
-  // We may eventually want to track preconnect_attempts to allow more preconnecting for
-  // heavily weighted upstreams or sticky picks.
-  if (shouldConnect(pending_streams_.size(), num_active_streams_, connecting_stream_capacity_,
-                    global_preconnect_ratio, true)) {
-    return true;
+  // Determine if we are trying to prefetch for global preconnect or local preconnect.
+  if (global_preconnect_ratio != 0) {
+    // If global preconnecting is on, and this connection is within the global
+    // preconnect limit, preconnect.
+    // For global preconnect, we anticipate an incoming stream to this pool, since it is
+    // prefetching for the next upcoming stream, which will likely be assigned to this pool.
+    // We may eventually want to track preconnect_attempts to allow more preconnecting for
+    // heavily weighted upstreams or sticky picks.
+    return shouldConnect(pending_streams_.size(), num_active_streams_, connecting_stream_capacity_,
```

Comment:
In a future change you might consider renaming connecting_stream_capacity_ also to make it match connecting and connected but no need to block this change on that.